### PR TITLE
Improve sampling errors for LUT

### PR DESCRIPTION
**Pull request by bjin**
the third commit is optional (feel free to ignore it)

### CODE REVIEW

**Comment by wm4** [comment]
Could LUT_POS not be a macro (taking the LUT size), instead of being redefined all over the place? It could be inserted inline by a function that generates shader code, too.

Intuitively, the LUT_POS calculation seems to be correct (it literally interpolates between the center of the first and last pixel in the LUT in the range ``[0-1]``). @haasn: what do you say?

I don't really see any need the make this configurable.

**Comment by bjin** [comment]
> Could LUT_POS not be a macro (taking the LUT size), instead of being redefined all over the place? It could be inserted inline by a function that generates shader code, too.

Not sure how to do that, that best I can think of is define it as function which returns the shader code as `const char*`, but the memory management gonna be messy.

> I don't really see any need the make this configurable.

Yes, for now I don't feel strongly about this either. But I'm considering some performance improvement to polar scaler (to get rid of all the `length()` call in shader), which make the coordinate of lookup texture non-linear. That will require a much larger texture size to make the sampling reasonably accurate. But again, it's not a issue for now and we can use different texture size for polar filter anyway, if necessary.

**Comment by wm4** [comment]
> Not sure how to do that, that best I can think of is define it as function which returns the shader code as const char*, but the memory management gonna be messy.

It doesn't need to be returned as string - it can just be directly added to the shader. Or this really could be just a macro that is defined once and takes the LUT size and "function" position as input. I wonder how you could avoid defining it twice (without the need for undef) - maybe you could just add it to the general shader prelude in gl_utils.c (where the compatibility macros etc. are defined). While this sounds slightly inelegant, I don't think this would cause any problems.

**Comment by bjin** [comment]
> maybe you could just add it to the general shader prelude in gl_utils.c (where the compatibility macros etc. are defined). 

The logic around LUT is too specific and I don't like the idea to move it into a generic place like `gl_util.c`, I added `lut_size` as the second parameter to `LUT_POS` to make it generic to be defined in `sampler_prelude()`.

**Comment by wm4** [comment]
I'm fine with the code. Will wait for haasn's comment.

**Comment by haasn** [comment]
This is a 6% slowdown. Is it equatable to a 6% improvement in accuracy?

**Comment by bjin** [comment]
> This is a 6% slowdown. 

Is it for polar scaler or just 1D scaler? How did you make the measurement?

**Comment by bjin** [comment]
Olay, get the context from IRC.

Pushed a new commit to make LUT texture size configurable. I didn't add an option to disable accurate lookup since
1. The performance issue is for polar scaler only.
2. Intuitively it's more accurate to decrease the texture size a bit, than disable accurate lookup.

**Comment by wm4** [comment]
I want to avoid: slow downs, slight incorrectness, new options. But it looks like we can't have all of them at the same time. Still, I think the LUT fixup should be included in _some_ form (it's 100% correct, and the code bjin wrote for it is neat too). Should it be another option?

**Comment by bjin** [comment]
With the new accurate lookup routines, there is actually little reason to use large texture size. I actually have a feeling what we can safely decrease the default texture size from 256 to some value like 64. With accurate coordinate, the bilinear intepolation should work reasonably well for any 1/63 interval of those continous filter kernels.

So actually I don't really see a reason to allow disable `LUT_POS`

**Comment by bjin** [comment]
Tried to estimate the lookup error, the code can be found at https://github.com/bjin/mpv/commit/d5b18d6d5822842cd47110264feb6a3282787304

* `accurate=true|false` means with or without `LUT_POS`
* Relative error is used since it's directly reflected on the final scaled images.
* Average error, standard error and maximum error are displayed in percentage
* Samples are uniformedly choosed from given radius

~~~
kernel=triangle, lut_size=256, accurate=false: avg=1.09346% std=5.61760% max=100.00000%
kernel=triangle, lut_size=256, accurate=true: avg=0.00000% std=0.00000% max=0.00000%
kernel=triangle, lut_size=64, accurate=true: avg=0.00000% std=0.00000% max=0.00000%
kernel=triangle, lut_size=16, accurate=true: avg=0.00000% std=0.00000% max=0.00000%
kernel=triangle, lut_size=128, accurate=true: avg=0.00000% std=0.00000% max=0.00000%
kernel=triangle, lut_size=1024, accurate=true: avg=0.00000% std=0.00000% max=0.00000%

kernel=bicubic, lut_size=256, accurate=false: avg=0.75495% std=1.32622% max=5.11709%
kernel=bicubic, lut_size=256, accurate=true: avg=0.01163% std=0.02508% max=0.13573%
kernel=bicubic, lut_size=64, accurate=true: avg=0.19010% std=0.40944% max=2.20009%
kernel=bicubic, lut_size=16, accurate=true: avg=3.14909% std=6.46380% max=29.15452%
kernel=bicubic, lut_size=128, accurate=true: avg=0.04688% std=0.10110% max=0.56145%
kernel=bicubic, lut_size=1024, accurate=true: avg=0.00072% std=0.00156% max=0.00864%

kernel=spline36, lut_size=256, accurate=false: avg=2.17348% std=6.28328% max=156.54582%
kernel=spline36, lut_size=256, accurate=true: avg=0.03137% std=0.11554% max=1.79150%
kernel=spline36, lut_size=64, accurate=true: avg=0.43740% std=1.10175% max=7.85595%
kernel=spline36, lut_size=16, accurate=true: avg=5.26964% std=8.61998% max=31.52875%
kernel=spline36, lut_size=128, accurate=true: avg=0.32378% std=2.64281% max=46.39838%
kernel=spline36, lut_size=1024, accurate=true: avg=0.00205% std=0.00965% max=0.28743%

kernel=mitchell, lut_size=256, accurate=false: avg=1.12743% std=2.38752% max=15.61202%
kernel=mitchell, lut_size=256, accurate=true: avg=0.02826% std=0.08851% max=1.02825%
kernel=mitchell, lut_size=64, accurate=true: avg=0.43858% std=1.24485% max=7.93260%
kernel=mitchell, lut_size=16, accurate=true: avg=6.43120% std=17.49451% max=199.63052%
kernel=mitchell, lut_size=128, accurate=true: avg=0.11520% std=0.37296% max=4.14150%
kernel=mitchell, lut_size=1024, accurate=true: avg=0.00176% std=0.00551% max=0.06394%

kernel=sinc, lut_size=256, accurate=false: avg=1.19984% std=5.63083% max=100.00000%
kernel=sinc, lut_size=256, accurate=true: avg=0.00982% std=0.06105% max=1.53785%
kernel=sinc, lut_size=64, accurate=true: avg=0.16559% std=1.05843% max=25.19183%
kernel=sinc, lut_size=16, accurate=true: avg=2.29812% std=9.19386% max=199.19535%
kernel=sinc, lut_size=128, accurate=true: avg=0.04046% std=0.25934% max=6.19980%
kernel=sinc, lut_size=1024, accurate=true: avg=0.00058% std=0.00312% max=0.09553%
~~~

So basically, lut_size=64 with LUT_POS is more accurate than lut_size=256 without LUT_POS. I will push another commit to decrease the default texture size down to 64 to compesate for the performance loss.

**Comment by bjin** [comment]
The same, but use absolute error instead.

~~~
kernel=triangle, lut_size=256, accurate=false: avg=0.000976578 std=0.001127664 max=0.001960754
kernel=triangle, lut_size=256, accurate=true: avg=0.000000000 std=0.000000000 max=0.000000000
kernel=triangle, lut_size=64, accurate=true: avg=0.000000000 std=0.000000000 max=0.000000000
kernel=triangle, lut_size=16, accurate=true: avg=0.000000000 std=0.000000000 max=0.000000000
kernel=triangle, lut_size=128, accurate=true: avg=0.000000000 std=0.000000000 max=0.000000000
kernel=triangle, lut_size=1024, accurate=true: avg=0.000000000 std=0.000000000 max=0.000000000

kernel=bicubic, lut_size=256, accurate=false: avg=0.000488321 std=0.000662963 max=0.001326985
kernel=bicubic, lut_size=256, accurate=true: avg=0.000003418 std=0.000004585 max=0.000015288
kernel=bicubic, lut_size=64, accurate=true: avg=0.000055995 std=0.000075103 max=0.000245958
kernel=bicubic, lut_size=16, accurate=true: avg=0.000989300 std=0.001320590 max=0.004001371
kernel=bicubic, lut_size=128, accurate=true: avg=0.000013777 std=0.000018484 max=0.000061268
kernel=bicubic, lut_size=1024, accurate=true: avg=0.000000212 std=0.000000285 max=0.000000954

kernel=spline36, lut_size=256, accurate=false: avg=0.001378340 std=0.002263460 max=0.005153949
kernel=spline36, lut_size=256, accurate=true: avg=0.000012053 std=0.000018815 max=0.000074277
kernel=spline36, lut_size=64, accurate=true: avg=0.000197358 std=0.000308009 max=0.001180920
kernel=spline36, lut_size=16, accurate=true: avg=0.003481978 std=0.005357684 max=0.018148406
kernel=spline36, lut_size=128, accurate=true: avg=0.000051366 std=0.000082913 max=0.000463984
kernel=spline36, lut_size=1024, accurate=true: avg=0.000000749 std=0.000001169 max=0.000004648

kernel=mitchell, lut_size=256, accurate=false: avg=0.000809448 std=0.001194892 max=0.002486225
kernel=mitchell, lut_size=256, accurate=true: avg=0.000006347 std=0.000008909 max=0.000030546
kernel=mitchell, lut_size=64, accurate=true: avg=0.000103993 std=0.000145780 max=0.000489919
kernel=mitchell, lut_size=16, accurate=true: avg=0.001829904 std=0.002551943 max=0.007855653
kernel=mitchell, lut_size=128, accurate=true: avg=0.000025587 std=0.000035903 max=0.000122292
kernel=mitchell, lut_size=1024, accurate=true: avg=0.000000394 std=0.000000554 max=0.000001907

kernel=sinc, lut_size=256, accurate=false: avg=0.001193036 std=0.001469751 max=0.002646229
kernel=sinc, lut_size=256, accurate=true: avg=0.000008449 std=0.000010639 max=0.000025296
kernel=sinc, lut_size=64, accurate=true: avg=0.000138405 std=0.000174282 max=0.000414085
kernel=sinc, lut_size=16, accurate=true: avg=0.002444195 std=0.003066796 max=0.007199235
kernel=sinc, lut_size=128, accurate=true: avg=0.000034066 std=0.000042892 max=0.000101964
kernel=sinc, lut_size=1024, accurate=true: avg=0.000000525 std=0.000000661 max=0.000001572
~~~

**Comment by haasn** [comment]
Performance figures. Obtained with

    apitrace trace -o mpv.trace ./build/mpv --no-config -no-audio --fullscreen clip.mp4 -end 0:10 -vo opengl-hq:scale=ewa_lanczossharp:cscale=ewa_lanczossharp:scaler-lut-size=N

followed by `qapitrace trace mpv.trace` and pressing (Ctrl+P) begin profiling. (There is some extra exercise in figuring out which of the program calls corresponds to which pass, but for scale and cscale it's obvious because they're the two slowest by far)

    lut=256, accurate: scale=5733μs, cscale=1581μs, total=7314μs
    lut=256, accurate: scale=5788μs, cscale=1597μs, total=7385μs
    lut=256, accurate: scale=5846μs, cscale=1611μs, total=7457μs

    lut=128, accurate: scale=5189μs, cscale=1452μs, total=6641μs
    lut=128, accurate: scale=5175μs, cscale=1449μs, total=6624μs
    lut=128, accurate: scale=4920μs, cscale=1378μs, total=6298μs

    lut=64, accurate: scale=4796μs, cscale=1344μs, total=6140μs
    lut=64, accurate: scale=4803μs, cscale=1368μs, total=6171μs
    lut=64, accurate: sacle=4646μs, cscale=1302μs, total=5948​μs

    lut=32, accurate: scale=4905μs, cscale=1374μs, total=6279μs
    lut=32, accurate: scale=4649μs, cscale=1303μs, total=5952μs
    lut=32, accurate: scale=4753μs, cscale=1331μs, total=6084μs

    lut=16, accurate: scale=5006μs, cscale=1404μs, total=6410μs
    lut=16, accurate: scale=4753μs, cscale=1332μs, total=6085μs
    lut=16, accurate: scale=4829μs, cscale=1352μs, total=6181μs

What I can see from these numbers is that it's not entirely clear that there's an improvement in speed at all for LUT sizes <64, probably because that size fits entirely into some sort of cache on my machine, but there's a relatively clear and relatively marked improvement in performance for the drop from 256 to 64.

As a reminder, here are the numbers for non-accurate lookup:

    lut=256, non-accurate, scale=5543μs, cscale=1533μs, total=7076μs
    lut=256, non-accurate, scale=5659μs, cscale=1562μs, total=7221μs
    lut=256, non-accurate, scale=5705μs, cscale=1575μs, total=7280μs

(Incidentally, in this test, the difference was more like 3%, though the exact figures vary depending on the clip and number of tests. 3 is almost certainly not enough tests to accurately characterize the performance since the spread is rather large from test to test)

Finally, I did some image comparisons. As expected, it's extremely hard to see any difference, but for the sake of placebo science I zoomed in as far as I could and squinted at pixels.

In most samples I looked at, it was extremely hard to see a difference at all at anything >=32, and usually I could only distinguish 16 samples from anything >16.

In some samples, I could “see” a difference all the way up to 64 vs 128 (accurate), although 256-nonaccurate looked more like 64 accurate.

Given these results, I agree with merging this code and making the default 64 in all cases. (I don't see a point in increasing it for opengl-hq, the speed drop is not worth it)

Even though there's little point in going past 64, keeping it configurable for speed reasons is justifiable since I don't know the performance characteristics of hardware other than my own.

**Comment by haasn** [comment]
Note that I think the absolute error is probably more interesting than the relative error because how much the error matters ultimately depends on the sample weight of that sample.

Most of the information in a pixel is coming from the primary few samples, and those are the ones with high sample values. An error there is going to have a bigger impact on the image than an error elsewhere.

**Comment by bjin** [comment]
The default scaler for `opengl-hq` is `spline36`, could you benchmark against the default setting?

I wouldn't expect huge performance drop for using `opengl-hq` default settings alone, but if there is a performance issue with polar scalers, the user can always decrease `scaler-lut-size` separately.

**Comment by haasn** [comment]
Note that this speed loss is only consequential for polar filters because polar filters are the only ones that query from the LUT many times per pixel.

It's possible this design could be changed, as I hypothesized on previous occasions via IRC:

Instead of going through N*N samples and querying sample(pos) * sample(LUT), adding them all together, then dividing by the total LUT amount, you could possibly instead use the same approach that is currently used for non-polar filters, but with a 3D LUT instead:

Query the LUT at vec3(fcoord, x) for x ranging from 0 to N^2/4, and save all the values in an array. Go through the N*N sample positions in a loop and look at the value of the LUT at that position. If it's above some critical threshold, sample from the pixel at that position.

This would allow for a 4x speedup (at minimum) in the sampling of the LUT, and fewer calculations in the loop. Probably also better cache behavior. It could be fairly significant, and allow us furthermore to offload the LUT fixup calculation to “once per pixel”.

**Comment by haasn** [comment]
> The default scaler for opengl-hq is spline36, could you benchmark against the default setting?

There's no difference here. It's a matter of individual microseconds. The error in the test is too big to even see a trend.

**Comment by wm4** [comment]
All in all, it sounds like this could be merged.

**Comment by wm4** [comment]
Merged, thanks.
